### PR TITLE
Always dispatch redraw events sequentially, one at a time

### PR DESCRIPTION
**Pull request by Lyude**
### TODO(?)

Do you need me to write any tests for this? Or do you think this is fine as it is?

---

After converting a fork of neovim-gtk that I've been working on over to
using nvim-rs, I noticed some rather strange issues with redrawing. In
particular, initiating completion in cmdline mode would rarely end up
resulting in the status bar being over-written by a line of completions,
followed by a blank line (which should have contained the command being
typed) all the way at the bottom of nvim's screen.

```
   _________________________________________
  |                                         | <- status line missing
  | foo/ bar/ baz/ biz/ boz/ bogz/ poggers/ |
  | fortnite/ memes/ runningoutofnames/     |
  |_|_______________________________________| <- should be :e
  ```

After some investigation, I noticed that nvim-rs dispatches all of its
requests asynchronously. This works fine for most things, with the
particular exception of UI redraw requests. Because each UI redraw request
assumes the state of the screen is equal to the result of the previous
redraw request, the client needs to have the guarantee that the redraw
requests coming in are always coming in the order they were originally sent
out by nvim. Since we currently don't expose the msgid of each notification
to our Handler, not maintaining that order means it's impossible for the
Handler to ascertain the correct order in which to perform each redraw.

So, let's fix this by implementing a queue for redraw requests, which we
use to flush redraw notifications from Neovim sequentially to the client in
parallel with all other RPC notifications.

**Licensing**: The code contributed to nvim-rs is licensed under the MIT or
Apache license as given in the project root directory.

### CODE REVIEW

**Comment by Lyude** [comment]
(it appears 1m is probably a bit too short of a timeout for tests :P)

File: src/neovim.rs [hunk 1]
```diff
@@ -44,6 +45,69 @@ type ResponseResult = Result<Result<Value, Value>, Arc<DecodeError>>;
...
+  if guard.is_active {
+    return;
+  }
+  guard.is_active = true;
+  drop(guard);
+
+  let queue = queue.clone();
+  let handler_c = handler.clone();
+  let neovim = nvim.clone();
+  handler.spawn(async move {
```
Comment (KillTheMule) [inline src/neovim.rs:91]: This probably needs a `+ create::Spawner` bound, so that you can actually use `handler::spawn`.

**Comment by KillTheMule** [comment]
Hey thanks! 

Actually, it's not just a problem with redraw notifications... *all* notifications and requests need to be handled in the correct sequence. There's only one fundamental problem, which was what prompted me to create nvim-rs. Namely, while handling a request, we can't block, since handling the request might trigger more communications before it's done (most simple example: in a request handler, try to get the current line... you need to be able to accept and deal with the response).

I've wrapped my head about this quite a bit, but haven't quite gotten it right. You can see my tries in the [serial](https://github.com/KillTheMule/nvim-rs/tree/serial) branch if you want. Do you see a way doing this right along the lines you've implemented it here? I'd rather solve this for all notifications/request and not redraws as a special case...
 
As for CI, I don't think a timeout is involved, but rather a forgotten trait bound. Not sure how the windows CI passed, though, need to investigate.

**Comment by Lyude** [comment]
> Hey thanks!
> 
> Actually, it's not just a problem with redraw notifications... _all_ notifications and requests need to be handled in the correct sequence. There's only one fundamental problem, which was what prompted me to create nvim-rs. Namely, while handling a request, we can't block, since handling the request might trigger more communications before it's done (most simple example: in a request handler, try to get the current line... you need to be able to accept and deal with the response).
> 
> I've wrapped my head about this quite a bit, but haven't quite gotten it right. You can see my tries in the [serial](https://github.com/KillTheMule/nvim-rs/tree/serial) branch if you want. Do you see a way doing this right along the lines you've implemented it here? I'd rather solve this for all notifications/request and not redraws as a special case...

A-ha! I had a suspicion this might need to be fixed for more then just UI events, as I was a bit surprised there was no ordering when I looked at the code.

Note that I glanced over the code you linked here but I'm not totally following what the idea here is, although maybe I just need to look over it again more closely at some point when I don't need to get back to my day job. From what I saw if I was understanding correctly, it seems like your idea might have been to block event delivery to all senders but the one handling a given method request? Or am I misunderstanding?

Either way, I think the strategy I'm using here (but maybe replace `RedrawQueue` with channels like you've used there, I'm quite new to rust so I didn't even realize these were a thing before :) could be changed around to work for what you're talking about.

Let's use a term from my day job here, timeline, to describe an indefinite sequence of events - e.g. methods, responses, and notifications - which must be presented to the client one after the other in order. I think my solution here would be to have two timelines: one specifically for method calls originating from nvim which we must respond to, and one for all other events. The idea here would be to ensure that getting responses from nvim is never blocked, which would allow us to handle methods safely. Likewise, the ordering we receive methods is guaranteed, and would happen in parallel with everything else.

I definitely will need to double check nvim's documentation on RPC when I get the time to see if there's any specifics on the expectations for how ordering with incoming methods and everything else correlate with each other. But my thinking here is that what's actually important is that we're receiving events from nvim chronologically, and that nvim receives responses from methods it's sent to us in the correct order. I don't _think_ it's an issue if we're handling these two timelines in parallel - actually I'd hope that nvim would have an expectation of methods it is sending to RPC clients being handled in parallel with notifications it is sending to wait clients.

Let me know what you think, and I'll be happy to try implementing this the next chance I get.

> 
> As for CI, I don't think a timeout is involved, but rather a forgotten trait bound. Not sure how the windows CI passed, though, need to investigate.

Ah - strange that it compiled on my system

**Comment by KillTheMule** [comment]
> but I'm not totally following what the idea here is

I totally understand, it _was_ a rough draft after all. There are 2 commits in there that try to solve the problem. The first consisted of custom polling the `handle_request` future, so we _do_ directly run it, but when it suspends and isn't making progress we would just handle the next incoming communication. There's a problem in there with the future being unexpectedly dropped (iirc it's the `poll!` that does it), but now that I'm rethinking it, this would not work out if the request handler suspends for another reason, i.e. waiting for it's own io.

The second idea was to just spawn of `handle_request`, giving it a clone of the writer, so it can finish business by itself without having to return a value. I don't know what the status of that is, I'll have to rethink what actually needs to be happening. You seem interested, so I'm going to ramble on a bit here ;)



>  The idea here would be to ensure that getting responses from nvim is never blocked, which would allow us to handle methods safely.

This isn't sufficient, though. We need to be able to handle requests and notifications while handling a request, too. See [this test](https://github.com/KillTheMule/nvim-rs/commit/e4d4010c9eea782c5235f88cab817dcbcb0889d7?branch=e4d4010c9eea782c5235f88cab817dcbcb0889d7&diff=unified#diff-d786987e1fca0e99fd36ffa17634b1708ef1545274a2e0aa1bfa574dd16b9c0e) as an artificial example of what might be happening.



> actually I'd hope that nvim would have an expectation of methods it is sending to RPC clients being handled in parallel with notifications it is sending to wait clients

That can't possibly work. Handlers can (and are expected to) modify some internal state of the plugin, and the order of operations needs to be controlled. Think e.g. of a plugin that does highlightin and reacts to edit events accordingly. If two notifications come in in fast succession, sending highlighting directives to neovim in the wrong order will wreak havoc.

### What's needed

I've made a headline here, because this sort of writing out what I'm thinking. First, some docs:

- https://github.com/msgpack-rpc/msgpack-rpc/blob/master/spec.md for the specification of the protocol (not the msgpack wire format, but the rpc format)
- https://github.com/neovim/pynvim The python client for neovim. Some things aren't specified properly, and the best answer to questions is "however pynvim doest it". I've not been able to fully understand its usage of greenlets (iirc it uses greenlets), though, but I guess I should give it another shot.

So here's how it needs to work out imho:

1. All events that arrive need to be handled in order, as I argued above
2. Exception: When a request arrives, we will need to handle all events that arrive during our handling. The reason is that neovim blocks while waiting for a response, but it's event loop keeps on running and everything that arrives during that is a direct consequence of our plugins' actions (e.g. responses to requests we make during that, but also requests might be nested, see the example test linked above).
3. (Maybe) exception from the exception: A notification arriving during our handling of a request might be unspecified behavior. It's not clear from the spec, and bfredl mentioned something along those lines some time ago.

So one consequence is that looping over incoming events, and then doing `await` on notification handlers does the right thing. On request handlers, that's probably not right, although I'm kinda thinking if we could pass the `Reader` into the request handling, so it could do it's own reading of incoming communications (recursively, as requests might nest). But that might mean we need to distinguish between "awaiting a response" or "awaiting  something else" in the request handler, which might not be doable.

The more I think on it, the more I feel like the last commit on `serial` is doing the right thing. Need to take that up... 

If you did, thanks for reading :) All thoughts welcome.

**Comment by Lyude** [comment]
> 
> If you did, thanks for reading :) All thoughts welcome.

I did! I think I definitely understand the problem much more now, and I definitely see why this was tricky to figure out :). If I get the time I'm going to look into trying to get that `serial` branch working as well, since it'd definitely be really useful to have this solved for my `neovim-gtk` work. Let me know if you get to it first though!

**Pull request closed without merging**